cli/src/main.rs

```diff
@@ -1,8 +1,89 @@
+use db::PgPool;
+use tokio;
+use clap::{Parser, Subcommand};
+use db::sqlx;
+use std::fs;
+use std::io::Write;
+use uuid::Uuid; // Ensure you have the `uuid` crate in your Cargo.toml
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct KnowledgeArgs {
+    #[command(subcommand)]
+    command: Command,
+}
+
+#[derive(Subcommand)]
+enum Command {
+    Login,
+    Signup,
+    DisplayUsers,
+}
+
 #[tokio::main]
 async fn main() -> color_eyre::Result<()> {
-    let _db_pool = db::setup_db_pool().await?;
+    let args = KnowledgeArgs::parse();
+    let db_pool = db::setup_db_pool().await?;
+
+    match args.command {
+        Command::Login => {
+            println!("Login not implemented yet")
+        },
+        Command::Signup => {
+            sign_up(&db_pool).await?;
+        },
+        Command::DisplayUsers => {
+            display_users(&db_pool).await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn persist_auth_session(user_id: Uuid) -> color_eyre::Result<()> {
+    if fs::metadata("auth.txt").is_ok() {
+        return Err(color_eyre::eyre::eyre!("auth.txt already exists"));
+    }
+
```

Comment:
We are doing this on purpose so we can make sure that user has signed up already

Reply:
This is great! Thanks for showing some examples of the error paths!

> The message Error: error returned from database: duplicate key value violates unique constraint "users_user_name_key is not coming from the persist_auth_session function

It might be nice to look into the `wrap_err` method that comes from `color_eyre` (well really from the `eyre` crate it uses, but close enough lol)
https://docs.rs/eyre/0.6.12/eyre/trait.WrapErr.html#tymethod.wrap_err
https://docs.rs/eyre/0.6.12/eyre/index.html#details

That will let you specify a more "custom" error message to also show the user! Your new "app specific" error message will be shown first, with the original error message being show as 'context' below it!